### PR TITLE
Improve EPA

### PR DESCRIPTION
**Pull request by lmontaut**
This PR is quite big, apologies for that in advance.
It does multiple things:
- It removes the successive mallocs in EPA when calling `GJKSolver`'s methods successively.
- To do so, it was necessary to refactor `GJKSolver`. Refactoring it removes copy-pastes between the different methods of the class, handling **all possible statuses** of GJK and EPA once they are called, adding documentation, adding logging to warn the user about problematic cases (only on debug mode or by activating the `HPP_FCL_ENABLE_LOGGING` preprocessor option).
- The fields `enable_contact` of `CollisionRequest` and `enable_signed_distance` of `DistanceRequest` now allow the user to make a trade-off between performance vs. contact information (sometimes we only want to know if two objects collide and only want to compute their distance when they are disjoint).
- It fixes the convergence criterion of EPA. The algo should be more robust now.

Because all of the statuses of GJK/EPA are now clear in the code, the truly degenerate cases of EPA (non-convex, invalid hull etc.) now trigger an assertion in debug mode. It turns out that some of the tests that we had do trigger these assertions (like `test/geometric_shapes.cpp` on cones or cylinders), which is good because it allows to understand the behavior of EPA. These tests are fixed now (EPA's tolerance cannot be too low on strictly-convex shapes, simply because it has terrible performance on them, which is expected from the algorithm).

### CODE REVIEW

File: include/hpp/fcl/narrowphase/gjk.h [hunk 1]
```diff
@@ -150,13 +150,25 @@ struct HPP_FCL_DLLAPI GJK {
...
+  /// class.
   struct HPP_FCL_DLLAPI Simplex {
     /// @brief simplex vertex
     SimplexV* vertex[4];
     /// @brief size of simplex (number of vertices)
     vertex_id_t rank;
 
     Simplex() {}
+
+    inline void reset() {
```
Comment (jorisv) [inline include/hpp/fcl/narrowphase/gjk.h:168]:
```suggestion
    void reset() {
```

Methods defined in a class/struct are inline by default.

File: src/narrowphase/gjk.cpp [hunk 2]
```diff
@@ -525,13 +525,17 @@ void MinkowskiDiff::set(const ShapeBase* shape0, const ShapeBase* shape1) {
...
   gjk_variant = GJKVariant::DefaultGJK;
   convergence_criterion = GJKConvergenceCriterion::VDB;
   convergence_criterion_type = GJKConvergenceCriterionType::Relative;
+  reset();
+}
+
+void GJK::reset() {
+  status = DidNotRun;
+  nfree = 0;
+  simplex = NULL;
```
Comment (jorisv) [inline src/narrowphase/gjk.cpp:538]:
```suggestion
  simplex = nullptr;
```

File: include/hpp/fcl/narrowphase/gjk.h [hunk 3]
```diff
@@ -379,15 +398,26 @@ struct HPP_FCL_DLLAPI EPA {
...
+  size_t iterations;
+
   Status status;
   GJK::Simplex result;
   Vec3f normal;
   FCL_REAL depth;
+
+ private:
   SimplexV* sv_store;
   SimplexF* fc_store;
```
Comment (jorisv) [inline include/hpp/fcl/narrowphase/gjk.h:416]: Why using raw ptr and managing memory manually ?

I think you can use safely a std::vector here.

File: include/hpp/fcl/narrowphase/gjk.h [hunk 4]
```diff
@@ -397,12 +427,32 @@ struct HPP_FCL_DLLAPI EPA {
...
+  /// @brief Copy constructor of EPA.
+  /// Mostly needed for the copy constructor of `GJKSolver`.
+  EPA(const EPA& other)
+      : max_face_num(other.max_face_num),
+        max_vertex_num(other.max_vertex_num),
+        max_iterations(other.max_iterations),
+        tolerance(other.tolerance) {
+    initialize();
+    memcpy(sv_store, other.sv_store, max_vertex_num * sizeof(SimplexV));
+    memcpy(fc_store, other.fc_store, max_face_num * sizeof(SimplexF));
```
Comment (jorisv) [inline include/hpp/fcl/narrowphase/gjk.h:439]: If you manage to turn sv_store and fc_store into a std::vector, you can use the default copy constructor.

File: src/CMakeLists.txt [hunk 5]
```diff
@@ -202,6 +202,12 @@ TARGET_LINK_LIBRARIES(${LIBRARY_NAME}
   Boost::filesystem
 )
 
+if (HPP_FCL_ENABLE_LOGGING)
+  TARGET_LINK_LIBRARIES(${LIBRARY_NAME} PUBLIC Boost::log)
+  # The compile flag `BOOST_LOG_DYN_LINK` is required here.
+  target_compile_definitions(${LIBRARY_NAME} PUBLIC -DHPP_FCL_ENABLE_LOGGING -DBOOST_LOG_DYN_LINK)
```
Comment (jorisv) [inline src/CMakeLists.txt:208]:
```suggestion
  target_compile_definitions(${LIBRARY_NAME} PUBLIC HPP_FCL_ENABLE_LOGGING BOOST_LOG_DYN_LINK)
```

File: include/hpp/fcl/narrowphase/gjk.h (hunk 3 above)
Comment (jorisv) [reply include/hpp/fcl/narrowphase/gjk.h:416]: If for some reason you can't use std::vector, try to use std::unique_ptr to add more safety.

**Review by jorisv — DISMISSED** [review]

File: include/hpp/fcl/narrowphase/gjk.h (hunk 3 above)
Comment (lmontaut) [reply include/hpp/fcl/narrowphase/gjk.h]: Using std::vector now. Thanks @jorisv !

File: CMakeLists.txt [hunk 6]
```diff
@@ -116,8 +117,14 @@ SET_BOOST_DEFAULT_OPTIONS()
 EXPORT_BOOST_DEFAULT_OPTIONS()
 IF(WIN32)
   ADD_PROJECT_DEPENDENCY(Boost REQUIRED COMPONENTS chrono thread date_time serialization filesystem)
+  if (HPP_FCL_ENABLE_LOGGING)
+    ADD_PROJECT_DEPENDENCY(Boost REQUIRED COMPONENTS log)
+  endif()
 ELSE(WIN32)
   ADD_PROJECT_DEPENDENCY(Boost REQUIRED chrono serialization filesystem)
+  if (HPP_FCL_ENABLE_LOGGING)
+    ADD_PROJECT_DEPENDENCY(Boost REQUIRED log)
```
[the remark this answers is not in the document]
Comment (lmontaut) [reply CMakeLists.txt]: Ok I will merge the two if statements and keep all the boost dependencies. Let see if it builds

[the remark this answers is not in the document]
Comment (lmontaut) [reply CMakeLists.txt]: Yep, seems to build  just fine on windows. Thanks!

**Pull request merged**